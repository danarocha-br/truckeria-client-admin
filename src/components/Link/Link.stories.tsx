import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import { Link, LinkProps } from '.';

export default {
  title: 'Navigation/Link',
  component: Link,
  parameters: {
    layout: 'centered',
    status: 'stable', // beta| deprecated | review | stable,
    design: {
      type: 'figma',
      url: '',
    },
  },
  args: {
    as: '',
    label: "I don't have an account",
    href: '/',
    variant: 'standalone',
  },
} as Meta;

const Template: Story<LinkProps> = (args) => <Link {...args} />;

export const Standalone = Template.bind({});
export const Inline = Template.bind({});
Inline.args = {
  variant: 'inline',
  label: 'Read our terms and conditions',
};
