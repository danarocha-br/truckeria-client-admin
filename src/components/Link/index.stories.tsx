import { Link } from '@chakra-ui/react';
import NextLink from 'next/link';
import { Story, Meta } from '@storybook/react/types-6-0';

export default {
  title: 'Components/Link',
  component: Link,
  argTypes: {
    variant: {
      control: {
        type: 'radio',
        options: ['primary'],
      },
    },
  },
  args: {
    variant: 'primary',
    isExternal: false,
  },
} as Meta;

const Template: Story = (args) => (
  <NextLink href="/" passHref>
    <Link {...args}>I am a Link</Link>
  </NextLink>
);

export const Primary = Template.bind({});
