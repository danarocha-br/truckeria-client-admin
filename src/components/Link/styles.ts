import { styled, keyframes } from '../../../stitches.config';
import { Clickable } from 'reakit/Clickable';

const lineUp = keyframes({
  '0%': {
    transformOrigin: '50% 100%',
    transform: 'scale3d(1, 0.045, 1)',
  },

  ' 50%': {
    transformOrigin: '50% 100%',
    transform: 'scale3d(1, 1, 1)',
  },

  '51%': {
    transformOrigin: '50% 0%',
    transform: 'scale3d(1, 1, 1)',
  },

  '100%': {
    transformOrigin: '50% 0%',
    transform: 'scale3d(1, 0.045, 1)',
  },
});

export const Anchor = styled(Clickable, {
  color: '$interactive-default',
  fontWeigth: '$medium',
  underline: 'none',
  d: 'inline-block',
  position: 'relative',
  pl: '$4',
  pr: '$5',
  transition: 'all cubic-bezier(0, 0, 0.2, 1) 700ms',
  transform: 'scale(1)',
  animationDirection: 'alternate-reverse',

  '&:after, &:before': {
    content: '',
    bg: '$action-primary-default',
    d: 'inline-block',
    w: '$5',
    h: '1.3px',
    position: 'absolute',
    top: '$3',
    transition: 'all cubic-bezier(0, 0, 0.2, 1) 700ms',
    transform: 'scale(0.85)',
  },

  '&:after': {
    ml: '$3',
  },

  '&:before': {
    ml: '-$7',
  },

  '&:hover': {
    color: '$interactive-hover',

    '&:before, &:after': {
      h: '1px',
      bg: '$action-primary-subdued',
      transform: 'scale(1.25)',
      transition: 'all cubic-bezier(0, 0, 0.2, 1) 700ms',
      animationDirection: 'alternate-reverse',
    },
  },
});

export const AnchorInline = styled(Clickable, {
  fontWeigth: '$medium',
  d: 'inline-block',
  position: 'relative',
  transition: 'all cubic-bezier(0, 0, 0.2, 1) 700ms',
  whiteSpace: 'nowrap',

  '&:hover': {
    color: '$interactive-hover',
  },

  variants: {
    appearance: {
      inline: {
        color: '$interactive-default',

        '&:before, &:after': {
          content: '',
          w: '$full',
          h: '1px',
          position: 'absolute',
          left: 0,
          pointerEvents: 'none',
          bg: 'currentColor',
          top: '100%',
        },

        '&:before': {
          opacity: 0,
          h: '3px',
        },

        '&:after': {
          opacity: 0,
          transitionProperty: 'opacity',
          transitionTimingFunction: 'cubic-bezier(0.4, 0, 0.2, 1)',
          transitionDuration: '150ms',
        },

        '&:hover': {
          '&:before': {
            opacity: 1,
            animation: `${lineUp} 0.3s ease forwards`,
          },

          '&:after': {
            opacity: 1,
            transitionDelay: '300ms',
          },
        },
      },

      'page-header': {
        color: '$text-default',
        fontSize: '$md',
        fontWeight: '$medium',
        textDecoration: 'none',
        opacity: 0.4,
        borderRadius: 2,

        outline: 'none',
        p: '$0',
        m: '$0',
        transitionProperty: 'opacity',
        transitionDuration: 300,

        '&:hover': {
          opacity: 1,
        },

        '&:focus': {
          outlineStyle: 'dashed',
          outlineOffset: 6,
          outlineWidth: 2,
          outlineColor: '$action-transparent-transparent',
          opacity: 1,
        },

        '@bp-md': {
          fontSize: '$lg',
        },
      },
    },

    isActive: {
      true: {},
    },
  },

  compoundVariants: [
    {
      isActive: true,
      appearance: 'page-header',
      css: {
        opacity: 1,
      },
    },
  ],

  defaultVariants: {
    appearance: 'inline',
  },
});
