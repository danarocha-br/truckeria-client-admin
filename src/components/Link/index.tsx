import * as React from 'react';
import { CSS } from '@stitches/react';
import NextLink from 'next/link';

import * as S from './styles';
import { config } from '../../../stitches.config';

export type LinkProps = {
  /** to render a wrapper to the a tag */
  as?: React.FunctionComponent | typeof NextLink;
  href: string;
  label?: string;
  /** if you need special link description */
  ariaLabel?: string;
  className?: string;
  variant?: 'standalone' | 'inline' | 'page-header';
  /** to be used with the variant page-header */
  isActive?: boolean;
  css?: CSS<typeof config>;
};

export const Link = ({
  href,
  as,
  label,
  ariaLabel,
  className,
  variant = 'standalone',
  isActive = false,
  css,
  ...props
}: LinkProps) => {
  const Component = as || React.Fragment;

  return (
    <Component href={href} passHref>
      {variant === 'standalone' ? (
        <S.Anchor
          as="a"
          className={className}
          aria-label={ariaLabel ? ariaLabel : label}
          css={css}
          {...props}
        >
          {label}
        </S.Anchor>
      ) : (
        <S.AnchorInline
          as="a"
          className={className}
          aria-label={ariaLabel ? ariaLabel : label}
          css={css}
          appearance={variant === 'page-header' ? 'page-header' : 'inline'}
          isActive={isActive}
          {...props}
        >
          {label}
        </S.AnchorInline>
      )}
    </Component>
  );
};
