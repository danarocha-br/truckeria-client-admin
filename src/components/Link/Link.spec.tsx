import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { renderWithTheme } from 'utils/tests/helpers';

import Link from '.';

describe('<Link />', () => {
  it('should render the Link', () => {
    const { container } = renderWithTheme(<Link to="/" label="I am a link" />);

    expect(screen.getByLabelText(/I am a link/i)).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('Is accessible by tab', () => {
    renderWithTheme(<Link to="/" label="I am a link" />);

    const link = screen.getByLabelText(/I am a link/i);
    expect(document.body).toHaveFocus();

    userEvent.tab();
    expect(link).toHaveFocus();
  });
});
