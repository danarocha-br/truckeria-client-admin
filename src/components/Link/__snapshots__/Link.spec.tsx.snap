// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> should render the Link 1`] = `
.c0 {
  --tw-translate-x: 0;
  --tw-translate-y: 0;
  --tw-rotate: 0;
  --tw-skew-x: 0;
  --tw-skew-y: 0;
  --tw-scale-x: 1;
  --tw-scale-y: 1;
  -webkit-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -ms-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform,filter,backdrop-filter;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  -webkit-transition-duration: 700ms;
  transition-duration: 700ms;
  font-size: 1.4rem;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: inline-block;
  position: relative;
  padding-left: 1rem;
  padding-right: 1.25rem;
  color: #CCAF81;
}

.c0:after,
.c0:before {
  position: absolute;
  display: inline-block;
  width: 1.25rem;
  top: 1.25rem;
  content: '';
  height: 1px;
  background-color: #FF8D4D;
}

.c0:after {
  margin-left: 0.75rem;
}

.c0:before {
  margin-left: -2rem;
}

.c0:hover {
  color: #FF8D4D;
}

.c0:hover:after,
.c0:hover:before {
  --tw-translate-x: 0;
  --tw-translate-y: 0;
  --tw-rotate: 0;
  --tw-skew-x: 0;
  --tw-skew-y: 0;
  --tw-scale-x: 1.25;
  --tw-scale-y: 1;
  -webkit-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -ms-transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  transform: translateX(var(--tw-translate-x)) translateY(var(--tw-translate-y)) rotate(var(--tw-rotate)) skewX(var(--tw-skew-x)) skewY(var(--tw-skew-y)) scaleX(var(--tw-scale-x)) scaleY(var(--tw-scale-y));
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,-webkit-transform,filter,backdrop-filter;
  -webkit-transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
  transition-property: background-color,border-color,color,fill,stroke,opacity,box-shadow,transform,filter,backdrop-filter;
  -webkit-transition-timing-function: cubic-bezier(0,0,0.2,1);
  transition-timing-function: cubic-bezier(0,0,0.2,1);
  -webkit-transition-duration: 700ms;
  transition-duration: 700ms;
}

<a
  aria-label="I am a link"
  class="c0"
  href="/"
>
  I am a link
</a>
`;
